js/drawobjects/pen.ts

```diff
@@ -9,10 +9,62 @@ that was still true after compressing with lzstring, with almost nothing drawn
 on the canvas and with lots of stuff drawn
 */
 
-import { Point, LineMode, DrawObject } from "./drawobject";
+import { Point, distance, LineMode, DrawObject } from "./drawobject";
 
 const POINT_DISTANCE_THRESHOLD: number = 2;
 
+const dotProduct = ([x1, y1]: Point, [x2, y2]: Point) => x1*x2 + y1*y2;
+const addVectors = ([x1, y1]: Point, [x2, y2]: Point) => [x1 + x2, y1 + y2] as [number, number];
+const subVectors = ([x1, y1]: Point, [x2, y2]: Point) => [x1 - x2, y1 - y2] as [number, number];
+
+/*
+The set [a1,b1] \ [a2,b2] is a union of 0, 1 or 2 intervals.
+This returns the end points of those intervals.
+Exported because tests.
+*/
+export function intervalSetDifference([a1, b1]: [number, number], [a2, b2]: [number, number]): [number, number][] {
+  return [
+    [a1, Math.min(a2, b1)] as [number, number],
+    [Math.max(a1, b2), b1] as [number, number],
+  ].filter(([start, end]) => start < end);
+}
+
+/*
+Return the parts of line segment between A and B that are outside a circle.
+If circle does not touch line segment at all, returns null.
+*/
+function splitLineSegmentWithCircle(A: Point, B: Point, center: Point, radius: number): [Point, Point][] | null {
+  // We represent the line as f(t) = A + (B-A)t, where 0 <= t <= 1
+  function f(t: number): Point {
+    return addVectors(A, subVectors(B, A).map(coordinate => coordinate*t) as [number, number]);
+  }
+
+  /*
+  Line and circle intersect when distance(f(t) - center) = radius
+  With dot products, we can rewrite that as dotProduct(f(t) - center, f(t) - center) = radius^2
+  After expanding, we can solve t with the quadratic formula
+  */
+  const AC = subVectors(A, center);
+  const BA = subVectors(B, A);
+  const ACBA = dotProduct(AC, BA);
+  const ACAC = dotProduct(AC, AC);
+  const BABA = dotProduct(BA, BA);
+
+  const squareRoot = Math.sqrt(ACBA*ACBA - ACAC*BABA + BABA*radius*radius);
+  if (isNaN(squareRoot)) {
+    // No intersection points of circle and line
+    return null;
+  }
+  const tMin = (-ACBA - squareRoot)/BABA;
+  const tMax = (-ACBA + squareRoot)/BABA;
+
+  const result = intervalSetDifference([0, 1], [tMin, tMax]).map(tInterval => tInterval.map(t => f(t)) as [Point, Point]);
+  if (result + "" == [[A, B]] + "") {
```

Comment:
Is there a better way of doing this than exploting JavaScript's weak typing?

Reply:
This code is comparing nested number arrays by value. It's intended to do this, but in a much less ugly way:

```js
if (result.length === 1 && result[0][0][0] === A[0] && result[0][0][1] === A[1] && result[0][1][0] === B[0] && result[0][1][1] === B[1]) {
```

What do you think about defining a `pointsEqual` function and then doing something like this?

```js
if (result.length === 1 && pointsEqual(result[0][0], A) && pointsEqual(result[0][1], B)) {
```